hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/TokenServiceContextImpl.java

```diff
@@ -18,84 +18,82 @@
+public class TokenServiceContextImpl implements StakingContext, FinalizeContext {
     private final Configuration configuration;
     private final ReadableStoreFactory readableStoreFactory;
     private final WritableStoreFactory writableStoreFactory;
+    private final RecordListBuilder recordListBuilder;
 
-    /**
-     * Constructs a {@link FinalizeContextImpl}.
-     *
-     * @param payer              The {@link AccountID} of the payer
-     * @param recordBuilder      The main {@link SingleTransactionRecordBuilderImpl}
-     * @param stack              The {@link SavepointStackImpl} used to manage savepoints
-     */
-    public FinalizeContextImpl(
-            @NonNull final AccountID payer,
-            @NonNull final SingleTransactionRecordBuilderImpl recordBuilder,
+    public TokenServiceContextImpl(
             @NonNull final Configuration configuration,
-            @NonNull final SavepointStackImpl stack) {
-        this.payer = requireNonNull(payer, "payer must not be null");
-        this.recordBuilder = requireNonNull(recordBuilder, "recordBuilder must not be null");
+            @NonNull final SavepointStackImpl stack,
+            @NonNull final RecordListBuilder recordListBuilder) {
         this.configuration = requireNonNull(configuration, "configuration must not be null");
+        this.recordListBuilder = requireNonNull(recordListBuilder, "recordListBuilder must not be null");
         requireNonNull(stack, "stack must not be null");
+
         this.readableStoreFactory = new ReadableStoreFactory(stack);
         this.writableStoreFactory = new WritableStoreFactory(stack, TokenService.NAME);
     }
 
-    @Override
-    @NonNull
-    public Instant consensusNow() {
-        return recordBuilder.consensusNow();
-    }
-
     @NonNull
     @Override
-    public AccountID payer() {
-        return payer;
+    public Instant consensusTime() {
+        return recordListBuilder.mainRecordBuilder().consensusNow();
     }
 
-    @Override
     @NonNull
+    @Override
     public Configuration configuration() {
         return configuration;
     }
 
-    @Override
     @NonNull
-    public <C> C readableStore(@NonNull final Class<C> storeInterface) {
+    @Override
+    public <T> T readableStore(@NonNull Class<T> storeInterface) {
         requireNonNull(storeInterface, "storeInterface must not be null");
         return readableStoreFactory.getStore(storeInterface);
     }
 
-    @Override
     @NonNull
-    public <C> C writableStore(@NonNull final Class<C> storeInterface) {
+    @Override
+    public <T> T writableStore(@NonNull Class<T> storeInterface) {
         requireNonNull(storeInterface, "storeInterface must not be null");
         return writableStoreFactory.getStore(storeInterface);
     }
 
-    @Override
     @NonNull
-    public <T> T recordBuilder(@NonNull final Class<T> recordBuilderClass) {
+    @Override
+    public <T> T mainRecordBuilder(@NonNull Class<T> recordBuilderClass) {
```

Comment:
Should this be `parentRecordBuilder` instead? I'm not totally sure I know exactly what a 'main' record is here